78: Meal plan

SakshiShreya opened the pull request:
Looks like this till now. I will update with every commit.

Mobile:
<img width="429" height="811" alt="image" src="https://github.com/user-attachments/assets/bfbc03c9-fddf-4674-807c-adbcbe053039" />

<img width="372" height="653" alt="image" src="https://github.com/user-attachments/assets/d085b163-c1ef-4ee2-9a91-c23864c74585" />

Desktop:
<img width="1440" height="811" alt="image" src="https://github.com/user-attachments/assets/f496eca5-a0c7-45f0-b9cc-37f083e9b1d0" />

<img width="1440" height="809" alt="image" src="https://github.com/user-attachments/assets/cf3b515f-9127-43f6-9d36-aea2e2bd7081" />

SakshiShreya committed:
79: Create meal plan blank screen

SakshiShreya committed:
82: Create a plan and redirect to plan page

SakshiShreya committed:
82: Increase plan name input width

Soumya-Shreya commented on src/planner/plans/edit/editPlanForm.tsx:
```diff
@@ -0,0 +1,49 @@
+});
+
+type Props = {
+  data: TCreatePlanBase;
```
Should't this be TPlan instead of TCreatePlanBase? You should be able to edit a plan with some already existing dishes as well

SakshiShreya replied:
Done

SakshiShreya committed:
Merge branch 'main' of https://github.com/SakshiShreya/Grocery-Planner into 79-meal-blank-screen

SakshiShreya committed:
78: Create new meal form

SakshiShreya committed:
79: Update TPlansBase to TPlans in plans response

SakshiShreya committed:
81: List all plans

SakshiShreya committed:
81: Fix suspended error on clicking browser back from edit plans screen

Van-sh commented on src/constants.ts:
```diff
@@ -26,3 +26,5 @@ export const emojis = [
   "😫",
 ];
+
+export const defaultUser = { id: "", name: "", fName: "", lName: "" };
```
```suggestion
export const defaultUser: TUser = { id: "", name: "", fName: "", lName: "" };
```

SakshiShreya replied:
Done

SakshiShreya committed:
79: Add mobile view

SakshiShreya committed:
79: Add desktop view

Van-sh commented on src/App.tsx:
```diff
@@ -32,6 +34,17 @@ function App() {
+                      element={
+                        <Suspense>
+                          <EditPlan />
+                        </Suspense>
```
move the `Suspense` to be the parent of the `Routes` component. All pages are lazy loaded so that should be the best catch-all suspense I think

SakshiShreya replied:
Yes. Looks like suspense in parent is better.

Done.

Van-sh commented on src/planner/plans/index.tsx:
```diff
@@ -0,0 +1,177 @@
+  } = useDisclosure();
+
+  const goToAddNewPlanPage = () => {
+    onCreateModalOpen();
+  };
+
+  const handleMutationSuccess = useCallback(
+    (action: string) => {
+      dispatch(
+        addToast({
+          message: `Plan ${action} successfully`,
+          type: "success",
+          autoClose: true,
+        }),
+      );
+      startTransition(() => {
+        navigate(`edit/${createData!.data._id}`);
+      });
+    },
+    [dispatch, navigate, createData],
+  );
```
This is clearly for createMutation if `createData` is accessed directly here. maybe should be renamed to handleCreateMutationSuccess and not take in action?

SakshiShreya replied:
Added condition for created action before navigation.
This function is used for delete also. I had not tested delete because api is not ready yet.

Van-sh commented on src/planner/plans/createForm.tsx:
```diff
@@ -0,0 +1,62 @@
+import { TCreatePlanBase } from "./types";
+
+export const schema = yup.object({
+  name: yup.string().required("Name is required"),
```
add `isPrivate` and `isActive` to the schema

SakshiShreya replied:
Done

Van-sh commented on src/planner/plans/createForm.tsx:
```diff
@@ -0,0 +1,62 @@
+      name: "",
+      isPrivate: false,
+      isActive: true,
+    } as TCreatePlanBase,
```
should be `satisfies` instead of `as` if it's supposed to do type checking, ignore if not

SakshiShreya replied:
I can use satisfies here. But if I do the same thing in addMeal.tsx, it gives error on certain places.

In addMeal.tsx, it thinks of dishes as never[].

I will have to write it like this there:
```js
initialValues: {
  mealType: "",
  dishes: [] as TMealDishBase[],
} satisfies TCreateMealBase,
```

Van-sh replied:
leave it as type cast then I think

Van-sh commented on src/planner/plans/createForm.tsx:
```diff
@@ -0,0 +1,62 @@
+          errorMessage={formik.errors.name}
+        />
+        <Checkbox {...formik.getFieldProps("isPrivate")}>Make Private</Checkbox>
+        <Checkbox {...formik.getFieldProps("isActive")} defaultSelected={formik.values.isActive}>
```
I doubt defaultSelected is required, `value` prop would already be passed because of `getFieldProps`

SakshiShreya replied:
This is required. Otherwise checkbox is not selected by default.

Van-sh commented on src/planner/plans/edit/types.ts:
```diff
@@ -0,0 +1,36 @@
+  "midafternoon" = "Mid Afternoon Snack",
+  "dinner" = "Dinner",
+  "bedtime" = "Bed Time",
+}
```
enums technically generate javascript code so I'm not sure I'm comfortable with this in types.ts.

have either the enum or the following in `constants.ts` 
```suggestion
export const mealTypes = {
  "wakeup": "Wake Up",
  "breakfast": "Breakfast",
  "midmorning": "Mid Morning",
  "lunch": "Lunch",
  "midafternoon": "Mid Afternoon Snack",
  "dinner": "Dinner",
  "bedtime": "Bed Time",
} as const
```

SakshiShreya replied:
We can use enums as both type and constant. That is why I am using enum.
I have moved the code to constant file but keeping enums as it is.

Van-sh commented on src/common/mealCards/index.tsx:
```diff
@@ -0,0 +1,33 @@
+
+export default function MealCards({ day }: Props) {
+  const { data } = useData();
+  const { meals = {} } = data;
```
we might need a better way to handle this if this is gonna be moved to common, or add a comment explaining the DataContext needs to be used
this would require moving the context to common as well I think

SakshiShreya replied:
Moved all related code to common. And added documentation comment also.

Van-sh commented on src/planner/plans/edit/addMeal.tsx:
```diff
@@ -0,0 +1,190 @@
+];
+
+const schema = yup.object({
+  mealType: yup.string().required("Meal type is required"),
```
add dishes

SakshiShreya replied:
Done

Van-sh commented on src/planner/plans/edit/addMeal.tsx:
```diff
@@ -0,0 +1,190 @@
+    initialValues: {
+      mealType: "",
+      dishes: [],
+      isPrivate: false,
```
remove isPrivate

SakshiShreya replied:
Done

Van-sh commented on src/planner/plans/edit/addMeal.tsx:
```diff
@@ -0,0 +1,190 @@
+            {({ push, remove }) => (
+              <>
+                {formik.values.dishes.map((_, index) => (
+                  <div key={index} className="bg-gray-100 flex flex-col gap-2 p-2 rounded-lg">
```
don't use index as the key

SakshiShreya replied:
Agreed.

We are using index in all forms. What shall we use instead?

One suggestion is to give timestamp to all new elements of array when we click on "Add dish"

Van-sh replied:
I think the dish's id should work?
that's what I decided to use in this PR https://github.com/Van-sh/Grocery-Planner/pull/95/changes#diff-2274da9f60ff6b7fa0ac10c309e13f75478d67e5d6fc9c3ba9b500f215156746R181

SakshiShreya replied:
What happens if we click on Add Dish? It will not have id until we select a dish.
We we click on Add Dish multiple times, then all of the blocks will have blank string.

Van-sh replied:
the add button is supposed to be disabled when any of the other values in array are not set
atleast that's how it works for ingredients in dish

SakshiShreya replied:
Ok then. I'll do the same here.

SakshiShreya replied:
Done

Van-sh reviewed (changes requested):
requested changes

SakshiShreya committed:
79: PR review updates

- Update yup schema
- Move mealcards and related code to common
- Add created contion before redirection from plan page
- Make all routes suspense

SakshiShreya committed:
79: Refetch on create/delete plan

Van-sh commented on src/planner/plans/edit/addMeal.tsx:
```diff
@@ -0,0 +1,201 @@
+                ))}
+                <Button
+                  variant="bordered"
+                  isDisabled={!!formik.getFieldMeta("ingredients").error}
```
this should be dishes

SakshiShreya replied:
Done

Van-sh commented on src/constants.ts:
```diff
@@ -27,4 +29,37 @@ export const emojis = [
+  midafternoon: EMealType.midafternoon,
+  dinner: EMealType.dinner,
+  bedtime: EMealType.bedtime,
+};
```
why is this needed?
I replaced it's only usage with EMealType and that seems to work just fine

SakshiShreya replied:
Yes. Replaced.

SakshiShreya committed:
79: Replace mealTypeMap with EMealType

SakshiShreya committed:
79: Fix validation and change key to non-index value

Van-sh reviewed (approved):
LGTM
